lightning/src/routing/network_graph.rs

```diff
@@ -668,6 +685,105 @@ impl_writeable_tlv_based!(ChannelInfo, {
 	(12, announcement_message, required),
 });
 
+/// A wrapper around [`ChannelInfo`] representing information about the channel as directed from a
+/// source node to a target node.
+pub struct DirectedChannelInfo<'a: 'b, 'b> {
+	channel: &'a ChannelInfo,
+	direction: Option<&'b DirectionalChannelInfo>,
+	source: &'b NodeId,
+	target: &'b NodeId,
+}
+
+impl<'a: 'b, 'b> DirectedChannelInfo<'a, 'b> {
+	/// Returns the node id for the source.
+	pub fn source(&self) -> &'b NodeId { self.source }
+
+	/// Returns the node id for the target.
+	pub fn target(&self) -> &'b NodeId { self.target }
+
+	/// Consumes the [`DirectedChannelInfo`], returning the wrapped parts.
+	pub fn into_parts(self) -> (&'a ChannelInfo, Option<&'b DirectionalChannelInfo>) {
+		(self.channel, self.direction)
+	}
+
+	/// Returns the [`EffectiveCapacity`] of the channel in a specific direction.
+	///
+	/// This is either the total capacity from the funding transaction, if known, or the
+	/// `htlc_maximum_msat` for the direction as advertised by the gossip network, if known,
+	/// whichever is smaller.
+	pub fn effective_capacity(&self) -> EffectiveCapacity {
+		Self::effective_capacity_from_parts(self.channel, self.direction)
+	}
+
+	/// Returns the [`EffectiveCapacity`] of the channel in the given direction.
+	///
+	/// See [`Self::effective_capacity`] for details.
+	pub fn effective_capacity_from_parts(
```

Comment:
We can drop this/merge into `effective_capacity` and drop `into_parts` if we just store a `DirectedChannelInfo` in `CandidateRouteHop::PublicHop`. Here's a sample, but you can do it by wrapping things instead of exposing the `channel` and `direction` fields in the struct if you want:

```diff
diff --git a/lightning/src/routing/network_graph.rs b/lightning/src/routing/network_graph.rs
index 1cb4f8606..2f9a1f35d 100644
--- a/lightning/src/routing/network_graph.rs
+++ b/lightning/src/routing/network_graph.rs
@@ -687,9 +687,10 @@ impl_writeable_tlv_based!(ChannelInfo, {
 
 /// A wrapper around [`ChannelInfo`] representing information about the channel as directed from a
 /// source node to a target node.
+#[derive(Clone, Debug)]
 pub struct DirectedChannelInfo<'a: 'b, 'b> {
-	channel: &'a ChannelInfo,
-	direction: Option<&'b DirectionalChannelInfo>,
+	pub(crate) channel: &'a ChannelInfo,
+	pub(crate) direction: Option<&'b DirectionalChannelInfo>,
 	source: &'b NodeId,
 	target: &'b NodeId,
 }
@@ -701,11 +702,6 @@ impl<'a: 'b, 'b> DirectedChannelInfo<'a, 'b> {
 	/// Returns the node id for the target.
 	pub fn target(&self) -> &'b NodeId { self.target }
 
-	/// Consumes the [`DirectedChannelInfo`], returning the wrapped parts.
-	pub fn into_parts(self) -> (&'a ChannelInfo, Option<&'b DirectionalChannelInfo>) {
-		(self.channel, self.direction)
-	}
-
 	/// Returns the [`EffectiveCapacity`] of the channel in a specific direction.
 	///
 	/// This is either the total capacity from the funding transaction, if known, or the
@@ -718,7 +714,7 @@ impl<'a: 'b, 'b> DirectedChannelInfo<'a, 'b> {
 	/// Returns the [`EffectiveCapacity`] of the channel in the given direction.
 	///
 	/// See [`Self::effective_capacity`] for details.
-	pub fn effective_capacity_from_parts(
+	fn effective_capacity_from_parts(
 		channel: &ChannelInfo, direction: Option<&DirectionalChannelInfo>
 	) -> EffectiveCapacity {
 		let capacity_msat = channel.capacity_sats.map(|capacity_sats| capacity_sats * 1000);
diff --git a/lightning/src/routing/router.rs b/lightning/src/routing/router.rs
index e6ad0faff..cb3f3e661 100644
--- a/lightning/src/routing/router.rs
+++ b/lightning/src/routing/router.rs
@@ -339,9 +339,8 @@ enum CandidateRouteHop<'a> {
 	},
 	/// A hop found in the [`NetworkGraph`], where the channel capacity may or may not be known.
 	PublicHop {
+		info: DirectedChannelInfo<'a, 'a>,
 		short_channel_id: u64,
-		channel: &'a ChannelInfo,
-		direction: &'a DirectionalChannelInfo,
 	},
 	/// A hop to the payee found in the payment invoice, though not necessarily a direct channel.
 	PrivateHop {
@@ -361,7 +360,7 @@ impl<'a> CandidateRouteHop<'a> {
 	fn features(&self) -> ChannelFeatures {
 		match self {
 			CandidateRouteHop::FirstHop { details } => details.counterparty.features.to_context(),
-			CandidateRouteHop::PublicHop { channel, .. } => channel.features.clone(),
+			CandidateRouteHop::PublicHop { info, .. } => info.channel.features.clone(),
 			CandidateRouteHop::PrivateHop { .. } => ChannelFeatures::empty(),
 		}
 	}
@@ -369,7 +368,7 @@ impl<'a> CandidateRouteHop<'a> {
 	fn cltv_expiry_delta(&self) -> u32 {
 		match self {
 			CandidateRouteHop::FirstHop { .. } => 0,
-			CandidateRouteHop::PublicHop { direction, .. } => direction.cltv_expiry_delta as u32,
+			CandidateRouteHop::PublicHop { info, .. } => info.direction.as_ref().unwrap().cltv_expiry_delta as u32,
 			CandidateRouteHop::PrivateHop { hint } => hint.cltv_expiry_delta as u32,
 		}
 	}
@@ -377,7 +376,7 @@ impl<'a> CandidateRouteHop<'a> {
 	fn htlc_minimum_msat(&self) -> u64 {
 		match self {
 			CandidateRouteHop::FirstHop { .. } => 0,
-			CandidateRouteHop::PublicHop { direction, .. } => direction.htlc_minimum_msat,
+			CandidateRouteHop::PublicHop { info, .. } => info.direction.as_ref().unwrap().htlc_minimum_msat,
 			CandidateRouteHop::PrivateHop { hint } => hint.htlc_minimum_msat.unwrap_or(0),
 		}
 	}
@@ -387,7 +386,7 @@ impl<'a> CandidateRouteHop<'a> {
 			CandidateRouteHop::FirstHop { .. } => RoutingFees {
 				base_msat: 0, proportional_millionths: 0,
 			},
-			CandidateRouteHop::PublicHop { direction, .. } => direction.fees,
+			CandidateRouteHop::PublicHop { info, .. } => info.direction.as_ref().unwrap().fees,
 			CandidateRouteHop::PrivateHop { hint } => hint.fees,
 		}
 	}
@@ -397,8 +396,8 @@ impl<'a> CandidateRouteHop<'a> {
 			CandidateRouteHop::FirstHop { details } => EffectiveCapacity::ExactLiquidity {
 				liquidity_msat: details.outbound_capacity_msat,
 			},
-			CandidateRouteHop::PublicHop { channel, direction, .. } => {
-				DirectedChannelInfo::effective_capacity_from_parts(channel, Some(direction))
+			CandidateRouteHop::PublicHop { info, .. } => {
+				info.effective_capacity()
 			},
 			CandidateRouteHop::PrivateHop { .. } => EffectiveCapacity::Infinite,
 		}
@@ -1054,16 +1053,12 @@ where L::Target: Logger {
 							let source = directed_channel.source();
 							let target = directed_channel.target();
 							if first_hops.is_none() || *source != our_node_id {
-								let (channel, direction) = directed_channel.into_parts();
-								if let Some(direction) = direction {
-									if direction.enabled {
+								if directed_channel.direction.is_some() && directed_channel.direction.as_ref().unwrap().enabled {
 										let candidate = CandidateRouteHop::PublicHop {
 											short_channel_id: *chan_id,
-											channel,
-											direction,
+											info: directed_channel,
 										};
 										add_entry!(candidate, *source, *target, $fee_to_target_msat, $next_hops_value_contribution, $next_hops_path_htlc_minimum_msat, $next_hops_path_penalty_msat);
-									}
 								}
 							}
 						}
@@ -1135,14 +1130,12 @@ where L::Target: Logger {
 					let target = NodeId::from_pubkey(&prev_hop_id);
 					let candidate = network_channels
 						.get(&hop.short_channel_id)
-						.and_then(|channel| channel.as_directed_to(&target).map(|d| d.into_parts()))
-						.and_then(|(channel, direction)| {
-							direction.map(|direction|
+						.and_then(|channel| channel.as_directed_to(&target))
+						.map(|info| {
 								CandidateRouteHop::PublicHop {
 									short_channel_id: hop.short_channel_id,
-									channel,
-									direction,
-								})
+									info,
+								}
 						})
 						.unwrap_or_else(|| CandidateRouteHop::PrivateHop { hint: hop });
 					let capacity_msat = candidate.effective_capacity().as_msat();
```

Reply:
Ah, that was simpler than I had originally thought.